### PR TITLE
Add Windows support (MSVC)

### PR DESCRIPTION
**Pull request by suhr**
NB: vcpkg-rs requires some environment variables to be set, and these variables cannot be set in `build.rs`.

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    VCPKG_TARGET: x86-windows
+    RUST_CHANNEL: nightly
+    LLVM_TARGET: win32
+
+install:
+  - ps: |
+      # Install Rust
+      appveyor-retry appveyor DownloadFile https://win.rustup.rs/ -FileName rustup-init.exe
+      .\rustup-init.exe -y --default-host "$env:TARGET" --default-toolchain "$env:RUST_CHANNEL"
+      $env:PATH += ";C:\Users\appveyor\.cargo\bin"
```
Comment (kpp) [inline appveyor.yml]: $env:Path += "$env:USERPROFILE\.cargo\bin"

File: appveyor.yml [hunk 2]
```diff
@@ -0,0 +1,61 @@
+image: Visual Studio 2017
+
+environment:
+  VCPKGRS_DYNAMIC: 1
+  RUST_CHANNEL: stable
+  LLVM_VERSION: 6.0.1
+  matrix:
+  - TARGET: x86_64-pc-windows-msvc
+    VCPKG_TARGET: x64-windows
+    LLVM_TARGET: win64
```
Comment (kpp) [inline appveyor.yml]: Missing `RUST_CHANNEL: stable`

File: appveyor.yml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  - TARGET: i686-pc-windows-msvc
+    VCPKG_TARGET: x86-windows
+    LLVM_TARGET: win32
```
Comment (kpp) [inline appveyor.yml]: Missing `RUST_CHANNEL: stable`

File: libsodium-sys/build.rs [hunk 4]
```diff
@@ -4,31 +4,25 @@ extern crate pkg_config;
...
-        let mode = match env::var_os("SODIUM_STATIC") {
-            Some(_) => "static",
-            None => "dylib",
-        };
-        if cfg!(target_env = "msvc") {
-            println!("cargo:rustc-link-lib={0}=libsodium", mode);
-        } else {
-            println!("cargo:rustc-link-lib={0}=sodium", mode);
-        }
+    let mode = match env::var_os("SODIUM_DYNAMIC") {
```
Comment (kpp) [inline libsodium-sys/build.rs]: Why did you change `SODIUM_STATIC` to `SODIUM_DYNAMIC`?

Comment (suhr) [reply libsodium-sys/build.rs]: I forgot to change it back actually.

**Comment by kpp** [comment]
Also change README.md "Platform Compatibiility -> Windows" "In Progress" to "Yes". Also it would be good to add badge. See example in tox-rs:

```
[![Appveyor Build Status][appveyor-badge]][appveyor-url] 

[appveyor-badge]: https://ci.appveyor.com/api/projects/status/y3y2hi6552qgmfr0/branch/master?svg=true
[appveyor-url]: https://ci.appveyor.com/project/kpp/tox/branch/master
```

And badge to Cargo.toml, example from tox-rs:
```
[badges]
appveyor = { repository = "kpp/tox", id = "y3y2hi6552qgmfr0" }
```

@Dylan-DPC give us the id of the project from AppveyorCI->Settings to setup the badge.

**Comment by suhr** [comment]
Should I also add a fallback (in the case when there's no libsodium in the system or required environment variables are not set) that downloads libsodium and links it statically?

**Comment by kpp** [comment]
> Should I also add a fallback

No. It's another issue. See https://github.com/sodiumoxide/sodiumoxide/issues/234

**Comment by Dylan-DPC** [comment]
Thanks a lot @suhr . I'm happy witih this. Is there anything else pending or (WIP) can be removed from the title and merged?

**Comment by suhr** [comment]
@Dylan-DPC It would be nice to add an appveyor badge.

**Comment by Dylan-DPC** [comment]
Ah damn forgot about that :D

**Comment by Dylan-DPC** [comment]
SVG image: https://ci.appveyor.com/api/projects/status/u05iy6wufw9ncdi7?svg=true

File: Cargo.toml [hunk 5]
```diff
@@ -20,6 +20,7 @@ members = ["libsodium-sys"]
 
 [badges]
 travis-ci = { repository = "sodiumoxide/sodiumoxide" }
+appveyor = { repository = "kpp/tox", id = "u05iy6wufw9ncdi7" }
```
Comment (kpp) [inline Cargo.toml]: kpp/tox =) Thanks

Comment (suhr) [reply Cargo.toml]: Oops.

File: appveyor.yml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      # Install Rust
+      appveyor-retry appveyor DownloadFile https://win.rustup.rs/ -FileName rustup-init.exe
+      cmd.exe /c .\rustup-init.exe -y --default-host "$env:TARGET" --default-toolchain "$env:RUST_CHANNEL" 2`>`&1
+      $env:PATH += ";$env:USERPROFILE\.cargo\bin"
+
+      # Install LLVM
+      echo "LLVM-${env:LLVM_VERSION}-${env:LLVM_TARGET}"
+      appveyor-retry appveyor DownloadFile "http://releases.llvm.org/${env:LLVM_VERSION}/LLVM-${env:LLVM_VERSION}-${env:LLVM_TARGET}.exe" -FileName llvm-installer.exe
+      7z x llvm-installer.exe -oc:\llvm-binary
+      #$env:PATH += ";C:\llvm-binary\bin"
```
Comment (kpp) [inline appveyor.yml]: Remove this comment

**Comment by kpp** [comment]
Although this build.rs passes CI, I don't like the algo behind it. It can be described in pseudolanguage:

```
find_libsodium():
1. If SODIUM_LIB_DIR is set, tell cargo to find library
2.1. If SODIUM_INC_DIR is set, return the value
2.2. Otherwise probe package using package manager and return the value of inc_dir
```

The point is that if the user specify any of env `SODIUM_LIB_DIR`, `SODIUM_INC_DIR`, `SODIUM_STATIC` we should use env vars only. We should use package managers only if there are no env variables set.

I don't want @suhr to implement correct algo since... well... we don't need this functionality. But in the future you guys will need to fix it.

**Comment by suhr** [comment]
> We should use package managers only if there are no env variables set.

I disagree about `SODIUM_STATIC`: on linux both static and dynamic libraries are available via `pkg-config`. It makes sense to link statically with a system libsodium.

**Comment by Dylan-DPC** [comment]
Will there be a case where CI doesn't have any of those values set? I know this should ideally mimic the way a developer builds the crate on windows, but we have other issues in that regard which we need to solve (such as #234 ), which should resolve this concern.

**Comment by kpp** [comment]
https://ci.appveyor.com/project/Dylan-DPC/sodiumoxide/builds/20147722/job/62pnocscycg3gm9h#L261

> warning: redundant linker flag specified for library `libsodium`

I wonder what's that

**Comment by Dylan-DPC** [comment]
bors: r+
